examples/android/clientcache/app/src/main/java/io/grpc/clientcacheexample/ClientCacheExampleActivity.java

```diff
@@ -0,0 +1,150 @@
+    mPortEdit = (EditText) findViewById(R.id.port_edit_text);
+    mMessageEdit = (EditText) findViewById(R.id.message_edit_text);
+    getCheckBox = (CheckBox) findViewById(R.id.get_checkbox);
+    noCacheCheckBox = (CheckBox) findViewById(R.id.no_cache_checkbox);
+    onlyIfCachedCheckBox = (CheckBox) findViewById(R.id.only_if_cached_checkbox);
+    mResultText = (TextView) findViewById(R.id.grpc_response_text);
+    mResultText.setMovementMethod(new ScrollingMovementMethod());
+
+    cache = SafeMethodCachingInterceptor.newLruCache(CACHE_SIZE_IN_BYTES);
+  }
+
+  public void sendMessage(View view) {
+    ((InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE))
+        .hideSoftInputFromWindow(mHostEdit.getWindowToken(), 0);
+    mSendButton.setEnabled(false);
+    new GrpcTask().execute();
+  }
+
+  private class GrpcTask extends AsyncTask<Void, Void, String> {
+    private String mHost;
+    private String mMessage;
+    private int mPort;
+    private ManagedChannel mChannel;
+
+    @Override
+    protected void onPreExecute() {
+      mHost = mHostEdit.getText().toString();
+      mMessage = mMessageEdit.getText().toString();
+      String portStr = mPortEdit.getText().toString();
+      mPort = TextUtils.isEmpty(portStr) ? 0 : Integer.valueOf(portStr);
+      mResultText.setText("");
+    }
+
+    @Override
+    protected String doInBackground(Void... nothing) {
+      try {
+        mChannel = ManagedChannelBuilder.forAddress(mHost, mPort).usePlaintext(true).build();
+        Channel channelToUse =
+            ClientInterceptors.intercept(
+                ManagedChannelBuilder.forAddress(mHost, mPort).usePlaintext(true).build(),
+                SafeMethodCachingInterceptor.newSafeMethodCachingInterceptor(cache));
+        HelloRequest message = HelloRequest.newBuilder().setName(mMessage).build();
+        HelloReply reply;
+        if (getCheckBox.isChecked()) {
+          MethodDescriptor<HelloRequest, HelloReply> safeCacheableUnaryCallMethod =
+              GreeterGrpc.METHOD_SAY_HELLO.toBuilder().setSafe(true).build();
+          CallOptions callOptions = CallOptions.DEFAULT;
+          if (noCacheCheckBox.isChecked()) {
+            callOptions =
+                callOptions.withOption(SafeMethodCachingInterceptor.NO_CACHE_CALL_OPTION, true);
+          }
+          if (onlyIfCachedCheckBox.isChecked()) {
+            callOptions =
+                callOptions.withOption(
+                    SafeMethodCachingInterceptor.ONLY_IF_CACHED_CALL_OPTION, true);
+          }
+          reply =
+              ClientCalls.blockingUnaryCall(
+                  channelToUse, safeCacheableUnaryCallMethod, callOptions, message);
+        } else {
+          GreeterGrpc.GreeterBlockingStub stub = GreeterGrpc.newBlockingStub(channelToUse);
+          reply = stub.sayHello(message);
+        }
+        return reply.getMessage();
+      } catch (Exception e) {
+        StringWriter sw = new StringWriter();
```

Comment:
Android uses LOG right?

Reply:
This is creating the string that gets displayed within the app, not writing to logcat. This is the approach used by both of our existing Android example apps: would you like to see something different here?

Reply:
Well, I would expect to see LOG, or let the exception propagate up.  I don't think the existing android examples were written with best practices.  

That said, I would have just expected the exception to get thrown.

Reply:
Added log. We'll get an exception here when e.g. the connection is UNAVAILABLE. This is printed to the screen, and now also logged.